### PR TITLE
Fix functional type related issues.

### LINKED ISSUE #25260: Error when defining variables of optional function types with no return types
**Description:**
$title.

**Steps to reproduce:**
```ballerina
function ()|() a = ();
```

```cmd
$ ballerina run qwe.bal 
Compiling source
        qwe.bal
error: .::qwe.bal:1:10: missing function name
error: .::qwe.bal:1:12: missing open brace token
error: .::qwe.bal:1:12: missing type desc
error: .::qwe.bal:1:23: missing close brace token
```

```ballerina
function ()? b = ();
```

```cmd
$ ballerina run qwe.bal 
Compiling source
        qwe.bal
error: .::qwe.bal:1:10: missing function name
error: .::qwe.bal:1:12: missing open brace token
error: .::qwe.bal:1:14: invalid token '?'
error: .::qwe.bal:1:21: missing close brace token
error: .::qwe.bal:1:14: undefined symbol 'b'
```

**Affected Versions:**
slp2

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
$Subject

Fixes #25260, #25259, #25287

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -992,11 +980,44 @@ private STNode parseReturnTypeDescRhs(STNode metadata, STNode functionKeyword, S
...
+                        typeDesc, fieldName);
+            } else {
+                return parseObjectFieldRhs(metadata, qualifiers.childInBucket(0), readonlyQualifier, typeDesc,
+                        fieldName);
+            }
+        }
+
+        STNode finalKeyword = STNodeFactory.createEmptyNode();
+        // Only the final keyword is allowed as a qualifier
+        for (int position = 0; position < qualifierList.size(); position++) {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1006]: do we need the similar validation for object members too (i.e: if `isObjectMember` is true) ?

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1006]: +1 will update.

File: compiler/ballerina-parser/src/main/resources/syntax_diagnostic_message.properties [hunk 2]
```diff
@@ -202,6 +215,8 @@ error.import.declaration.after.other.declarations=import declaration after other
 error.annotations.attached.to.expression=annotations attached to expression
 error.invalid.expression.in.start.action=invalid expression in start action
 error.same.object.type.qualifier=same object type qualifier
+error.duplicate.qualifier=duplicate qualifier
+error.qualifier.not.allowed=qualifier not allowed
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/resources/syntax_diagnostic_message.properties:219]: I think error msg like this:  `''{0}'' qualifier not allowed here` would be better. 
In the class-change branch it is already using that way for `error.qualifier.not.allowed`

File: compiler/ballerina-parser/src/main/resources/syntax_diagnostic_message.properties [hunk 3]
```diff
@@ -202,6 +215,8 @@ error.import.declaration.after.other.declarations=import declaration after other
 error.annotations.attached.to.expression=annotations attached to expression
 error.invalid.expression.in.start.action=invalid expression in start action
 error.same.object.type.qualifier=same object type qualifier
+error.duplicate.qualifier=duplicate qualifier
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/resources/syntax_diagnostic_message.properties:218]: `duplicate ''{0}'' qualifier` would be better

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -992,11 +980,42 @@ private STNode parseReturnTypeDescRhs(STNode metadata, STNode functionKeyword, S
...
+            STNode qualifier = qualifierList.get(position);
+            if (isObjectMember) {
+                // public or private qualifier allowed in object field.
+                if (qualifier.kind == SyntaxKind.PUBLIC_KEYWORD || qualifier.kind == SyntaxKind.PRIVATE_KEYWORD) {
+                    visibilityQualifier = qualifier;
+                } else {
+                    // if remote or transactional qualifier present
+                    typeDesc = SyntaxErrors.cloneWithLeadingInvalidNodeMinutiae(typeDesc, qualifier,
+                            DiagnosticErrorCode.ERROR_QUALIFIER_NOT_ALLOWED);
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1002]: Related above comment. 
```suggestion
                    typeDesc = SyntaxErrors.cloneWithLeadingInvalidNodeMinutiae(typeDesc, qualifier,
                            DiagnosticErrorCode.ERROR_QUALIFIER_NOT_ALLOWED, ((STToken) qualifier).text()));
```

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 5]
```diff
@@ -992,11 +980,42 @@ private STNode parseReturnTypeDescRhs(STNode metadata, STNode functionKeyword, S
...
+    private STNode parseVarDeclWithFunctionType(STNode typeDesc, boolean isObjectMember, STNode qualifiers,
+                                                STNode metadata) {
+        STNodeList qualifierList = (STNodeList) qualifiers;
+        STNode visibilityQualifier = STNodeFactory.createEmptyNode();
+        // qualifiers are not allowed except for object filed
+        for (int position = 0; position < qualifierList.size(); position++) {
+            STNode qualifier = qualifierList.get(position);
+            if (isObjectMember) {
+                // public or private qualifier allowed in object field.
+                if (qualifier.kind == SyntaxKind.PUBLIC_KEYWORD || qualifier.kind == SyntaxKind.PRIVATE_KEYWORD) {
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:997]: I think we can merge this check and avoid nested if clause.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 6]
```diff
@@ -992,11 +980,42 @@ private STNode parseReturnTypeDescRhs(STNode metadata, STNode functionKeyword, S
...
+            } else {
+                typeDesc = SyntaxErrors.cloneWithLeadingInvalidNodeMinutiae(typeDesc, qualifier,
+                        DiagnosticErrorCode.ERROR_QUALIFIER_NOT_ALLOWED);
+            }
+        }
+        if (isObjectMember) {
+            STNode readonlyQualifier = STNodeFactory.createEmptyNode();
+            STNode fieldName = parseVariableName();
+            return parseObjectFieldRhs(metadata, visibilityQualifier, readonlyQualifier, typeDesc, fieldName);
+        }
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1013]: new lines at top and bottom would be cleaner

File: compiler/ballerina-parser/src/main/resources/syntax_diagnostic_message.properties (hunk 3 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/resources/syntax_diagnostic_message.properties:218]: thanx will change

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 5 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: changed

**Review by SupunS — APPROVED** [review]

**Pull request merged**